### ISSUE TITLE
Verify azp claim 

### ISSUE THREAD

**Issue by manjinder-singh-tech** [coreos/go-oidc#355]
Is there a specific reason, verifier does not support verifying `azp` claim? It could be possible for some scenarios that client is explicitly required to verify that `azp` claim matches ClientID. Can we add support for verifying that by any chance, it does not have to be enforced and can be optional (via verifier config) to satisfy certain use cases for say custom providers.

**Comment by ericchiang** [comment]
There's not a strong reason. I've only seen azp used in relatively niche scenarios where you still want to verify the audience rather than authorizing party.

Do you have a use case in mind? What's the desired security outcome?

**Comment by manjinder-singh-tech** [comment]
We have a custom Identity Provider, that for some reason supports other entities as ‘aud’ rather than ClientID, and the authorized party sends ClientID as part of ‘azp’ in the token, and requires us to validate that instead of aud. 

I understand that perhaps this not a very common use case, but hoping it can still be beneficial to others. If we can provide optional config attribute to validate ‘azp’ to match ClientID.

**Comment by ericchiang** [comment]
Today you can pull the 'azp' claim out of the IDToken with the Claims() method and do your validation on that.

https://pkg.go.dev/github.com/coreos/go-oidc/v3/oidc#IDToken.Claims

It does appear that the azp claim is supposed to be validated after the token endpoint call

"If an azp (authorized party) Claim is present, the Client SHOULD verify that its client_id is the Claim Value."

https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation

The problem is that this package can be used by both the server requesting the token and other service that might want to validate the token later. For the former, you want to verify 'azp' matches your client_id. For the latter, you'd want to validate 'aud'.

So yeah, maybe we should have an "authorizing party" validation option?

**Comment by manjinder-singh-tech** [comment]
Makes sense i believe. We can make it optional to satisfy both server and client implementations, and the verifier config can be adjusted accordingly.
Aldo, thanks for suggesting the Claims method.

**Comment by ericchiang** [comment]
Another thought is that we could have a method like:

```
func (v *IDTokenVerifier) VerifyToken(ctx context.Context, token *oauth2.Token) (*IDToken, error)
```

That did the right thing with the authorizing party claim. Might be a bit subtle though... Let me think about this a little

**Comment by trixpan** [comment]
@ericchiang the spec states 

> 4. If an azp (authorized party) Claim is present, the Client SHOULD verify that its client_id is the Claim Value.

https://openid.net/specs/openid-connect-basic-1_0.html#IDTokenValidation

**Comment by 1seal** [comment]
i put together an opt-in implementation in #474.

it adds Config.VerifyAuthorizedParty (default false). when enabled:
- if aud has multiple values, azp must be present and equal to ClientID
- if azp is present, it must be equal to ClientID

the idea is to support deployments that want spec-aligned binding in the multi-audience case (and avoid accepting a token where ClientID is merely one element of aud while azp points at a different client), without changing default behavior for callers that validate tokens outside the original client/token-exchange context.

**Comment by ericchiang** [comment]
Thanks! However, as discussed in https://github.com/coreos/go-oidc/issues/355#issuecomment-1287435457, it's not clear that azp should always be equivalent to client ID. E.g. this package is used by servers to verify ID Tokens issued to other relying parties like Kubernetes which accepts ID Tokens as an authentication strategy. In that scenario, the whole point of azp it that it can be different than aud and allow different clients to request access on another's behalf.

In that scenario, the option proposed in #474 isn't correct. At the point where you have to reason about handling azp anyway, you can already do that today in the library by calling https://pkg.go.dev/github.com/coreos/go-oidc/v3/oidc#IDToken.Claims

I mentioned this is the private email thread, but I'm probably not going to make changes unless this actually has a negative interaction with a real provider. Being spec-aligned in the abstract isn't a goal of this library. Particularly when even OpenID Connect certified clients don't implement the logic you're describing:

- https://github.com/zitadel/oidc/blob/v3.45.4/pkg/oidc/verifier.go#L125
- https://github.com/panva/oauth4webapi/blob/0fc74562ed9d42604fcf9ce060ac2fddce146af5/src/index.ts#L3737

**Comment by 1seal** [comment]
thanks, understood.

i agree my proposed flag in #474 assumes a “post-token-endpoint client” context and is not generally correct for third-party verifiers (e.g. kubernetes) where azp semantics can legitimately differ.

i don’t currently have a real-world provider interaction to motivate expanding api surface here, and callers can already enforce their own azp policy via IDToken.Claims. i’ll close #474 to reduce noise.

if a docs-only pr would be welcome, i can add a short note/examples on when/how to validate azp (and why it is context-dependent), and point back to this issue.

**Comment by 1seal** [comment]
follow-up on the “real-world provider” point: dex documents cross-client trust where an ID Token is intentionally issued with aud set to a different client_id than the client that performed the login (and azp identifies the client that requested the token).

- https://dexidp.io/docs/configuration/custom-scopes-claims-clients/#cross-client-trust-and-authorized-party
- https://dexidp.io/docs/openid-connect/ (example shows aud as an array and azp set)

this matches the kubernetes oidc auth doc guidance that azp can be used to allow one client to issue tokens on behalf of another, and that the apiserver is not itself an oauth client:
- https://kubernetes.io/docs/reference/access-authn-authz/authentication/ (search for azp)

so i agree azp == “verifier client id” is not a generally correct rule for this library, and my earlier proposed flag in #474 was too opinionated for these deployments.

**Comment by ericchiang** [comment]
Yes, I'm familiar. I wrote both of those implementations
